src/FreeAge/client/game_controller.cpp

```diff
@@ -394,6 +397,50 @@ void GameController::HandleUnitMovementMessage(const QByteArray& data) {
   unit->SetMovementSegment(currentGameStepServerTime, startPoint, speed, action, map.get(), match.get());
 }
 
+void GameController::HandleUnitGarrisonMessage(const QByteArray& data) {
+  if (data.size() < 4) {
+    LOG(ERROR) << "Received a too short UnitGarrison message";
+    return;
+  }
+  const char* buffer = data.data();
+  
+  u32 unitId = mango::uload32(buffer + 0);
+  u32 targetObjectId = mango::uload32(buffer + 4);
+
+  auto unitIt = map->GetObjects().find(unitId);
+  if (unitIt == map->GetObjects().end()) {
+    LOG(ERROR) << "Received a UnitGarrison message for an object ID that is not in the map.";
+    return;
+  }
+  if (!unitIt->second->isUnit()) {
+    LOG(ERROR) << "Received a UnitGarrison message for an object ID that is a different type than a unit.";
+    return;
+  }
+  auto targetObjectIt = map->GetObjects().find(targetObjectId);
+  if (targetObjectIt == map->GetObjects().end()) {
+    LOG(ERROR) << "Received a UnitGarrison message with target object ID that is not in the map.";
+    return;
+  }
+
+  ClientUnit* unit = AsUnit(unitIt->second);
+  GarrisonUnit(unitId, unit, targetObjectId, targetObjectIt->second, !unit->IsGarrisoned());
+}
+
+void GameController::GarrisonUnit(u32 unitId, ClientUnit* unit, u32 targetObjectId, ClientObject* targetObject, bool enter) {
+  // TODO: implement a way to get ID from ClientObject, then remove the ids for the arguments
```

Comment:
Is there a reason that I can't why the ClientObject and ServerObject would not have a id field or it was just no need for it?

Reply:
So far there was no real need for it since I always got along with passing both the object pointer and the corresponding ID around. If you prefer having the ID within the object for convenience, I think that would be fine as well: the additional memory cost should be negligible and since the IDs are constant, there is no danger of them getting out of sync.